### ISSUE TITLE
`@final` no longer checked with the new `TypedDict` substitutability rule

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#2608]
## Environment

pyright from `main` (commit `7f4b0b0036f3b8ab928cf9d95bbef8fb10842be0`)

## Bug

After #1899 pyright made a new rule that because of subclassing, type narrowing on `TypedDict`s in a union should only work if they are all `@final`.

Solving #2592 introduced a regression in this regard. You can sneak in a subclass that accidentally has the same field as a different union member.

## Code example
```py
from typing import TypedDict, final, Union, List


@final
class A(TypedDict):
    foo: int
    fizz: str
        

@final
class B(TypedDict):
    bar: str


class C(TypedDict):
    bar: str


class D(C):
    foo: List[int]


def f(thing: Union[A, B]):
    if "foo" in thing:
        reveal_type(thing)
        print(thing["fizz"])
    else:
        reveal_type(thing)
        print("no fizz")


def g(c: C):
    f(c)


impostor: D = {"bar": "baz", "foo": [1, 2, 3]} 
g(impostor)
```

## What I assume is happenning

1. From `f`'s point of view, the type narrowing is legal because `A` and `B` are `@final`
2. From `g`'s point of view, `C` can be assigned to `Union[A, B]` because `C` can be assigned to `B` due to the new rule (i.e. they have the same fields).
3. From `g(impostor)`'s point of view, `D` can be assigned to `C` because it subclasses it

At runtime, this fails, because `f` will incorrectly decide that `impostor` is an `A` and access `"fizz"` on it.

To be clear, I don't have a real example or a use case where this is a problem. This is just a side effect of the new behaviour I found.

**Comment by decorator-factory** [comment]
Possible solution: non-`@final` `TypedDict` should not be assignable to a `@final` `TypedDict`?

**Comment by erictraut** [comment]
Yes, I agree that's the best solution. A fix will be included in the next version.

**Comment by erictraut** [comment]
This is included in pyright 1.1.190, which I just published. It will also be included in the next release of pylance.
